### ISSUE TITLE
Support types other than string for secondary indexes

### ISSUE THREAD

**Issue by aimeelaplant** [typedorm/typedorm#109]
Hello, thanks for making an awesome ORM.

Will there be support for specifying a type other than a string for the secondary indexes? DynamoDB allows you to pick numbers as a value for a secondary index partition keys and sort keys, so there are many existing designs with numbers.

So, something like this:

```
const myGlobalTable = new Table({
  name: 'test-table',
  partitionKey: 'PK',
  sortKey: 'SK',
  indexes: {
    GSI1: {
      type: INDEX_TYPE.GSI,
      partitionKey: 'GSI1PK',
      partitionValueType: 'number', // adding value type
      sortKey: 'GSI1SK',
    },
  },
});
```

In the entity mapping:

```
import {Attribute, Entity, AutoGeneratedAttribute} from '@typedorm/common';
import {AUTO_GENERATE_ATTRIBUTE_STRATEGY} from '@typedorm/common';

@Entity({
  name: 'organisation',
  primaryKey: {
    partitionKey: 'ORG#{{id}}',
    sortKey: 'ORG#{{id}}',
  },
  indexes: {
    GSI1: {
      partitionKey: '{{numberValue}}', // <-- gets converted to number type when it goes to Dynamo.
      sortKey: 'ORG#{{id}}#ACTIVE#{{active}}',
      type: INDEX_TYPE.GSI,
    },
  },
})
export class Organisation{

  @AutoGeneratedAttribute({
    strategy: AUTO_GENERATE_ATTRIBUTE_STRATEGY.UUID4,
  })
  id: string;

  @Attribute()
  numberValue: number;
}
```

Also, would there be support for converting attribute values to a different type from another library? Here is how the `class-transformer` library does it:

```
export class Photo {
  id: number;

  @Type(() => Date)
  @Transform(({ value }) => moment(value), { toClassOnly: true })
  date: Moment;
}
``` 

So something like this could work:

```
export class Organisation{
  @Attribute({ toDynamo: (value) => value.toString(), fromDynamo: (value: string) => DateTime.fromString(value))
  createdAt: DateTime;
}
```

**Comment by aimeelaplant** [comment]
@whimzyLive Makes sense! There are some examples on AWS that use numbers as global secondary indexes for write sharding. You can see an example here: https://amazon-dynamodb-labs.com/design-patterns/ex3gsisharding/step2.html. So unfortunately there are a lot of existing designs with numbers as gsi's.

> Not entirely sure about this rn, but I like the idea of being able to modify attribute value in `toDynamo` and `fromDynamo` hooks, will definitely look into this.

Great. I am not familiar with TypeORM though it uses the same data mapper pattern as Hibernate and Doctrine, and in those worlds, they use an AttributeValueConverter class.

**Comment by whimzyLive** [comment]
closing this as both requests are tracked as a separate issues, #112 and #113

**Comment by whimzyLive** [comment]
@aimeelaplant attribute aliasing is now available to try in 1.13.0-beta.0. Chechout [this](https://github.com/typedorm/typedorm/blob/beta/docs/how-to-recipes.md#define-an-entity-with-index).

**Comment by whimzyLive** [comment]
Support for attribute transform is available to play within 1.14.0-beta.2. Read more [here](https://github.com/typedorm/typedorm/blob/beta/docs/how-to-recipes.md#attribute-transformation)

**Comment by jarredkenny** [comment]
How can attribute transformations be used on an LSI or GSI if the LSISK or GSISK is generated internally by typedorm?

I have this use case:
```typescript
@Entity({
  name: "order",
  primaryKey: {
    partitionKey: "DOMAIN#{{domain}}",
    sortKey: "ORDER#{{id}}",
  },
  indexes: {
    LSI1: {
      type: INDEX_TYPE.LSI,
      sortKey: "{{date}}",
    },
    LSI2: {
      type: INDEX_TYPE.LSI,
      sortKey: "LOCATION#{{locationKey}}",
    },
  },
})
```
Where `LSI1` is `number` so that BETWEEN can be used on my queries. Right now, typeorm always tries to set `LSI1SK` as a `string` which throws an index error.
